src/hotspot/share/gc/shared/objectCountClosure.hpp

```diff
@@ -0,0 +1,29 @@
+#ifndef SHARE_GC_SHARED_OBJECTCOUNTCLOSURE_HPP
+#define SHARE_GC_SHARED_OBJECTCOUNTCLOSURE_HPP
+
+#include "gc/shared/gcId.hpp"
+#include "gc/shared/objectCountEventSender.hpp"
+#include "jfr/jfrEvents.hpp"
+#include "memory/allStatic.hpp"
+#include "runtime/mutex.hpp"
+#include "memory/heapInspection.hpp"
+#include "utilities/macros.hpp"
+#include "utilities/ticks.hpp"
+
+class KlassInfoEntry;
+class Klass;
+
+class ObjectCountClosure : public AllStatic {
+    static KlassInfoTable* cit;
+
+public:
+    static bool check_table_exists();
+    static bool record_object(oop o);
+    static KlassInfoTable* get_table();
+    static void reset_table();
+
+    template <class Event>
+    static bool should_send_event();
```

Comment:
I am not convinced you need all these methods to be static. This ties you up from using different closures writing, for example to their "local tables".

What would happen if you stuck the (global) klass info table in CollectedHeap, or in ShenandoahHeap, and seeded the closure with the klass info table it would use.

That would allow one to easily refactor in the future to use local tables in the future without needing wholesale refactoring.